packages/server/src/sdk/app/tables/getters.ts

```diff
@@ -0,0 +1,128 @@
+import { context } from "@budibase/backend-core"
+import {
+  BudibaseInternalDB,
+  getMultiIDParams,
+  getTableParams,
+} from "../../../db/utils"
+import {
+  breakExternalTableId,
+  isExternalTable,
+  isSQL,
+} from "../../../integrations/utils"
+import {
+  AllDocsResponse,
+  Database,
+  Table,
+  TableResponse,
+  TableViewsResponse,
+} from "@budibase/types"
+import datasources from "../datasources"
+import sdk from "../../../sdk"
+
+function processInternalTables(docs: AllDocsResponse<Table[]>): Table[] {
+  return docs.rows.map((tableDoc: any) => ({
+    ...tableDoc.doc,
+    type: "internal",
+    sourceId: tableDoc.doc.sourceId || BudibaseInternalDB._id,
+  }))
+}
+
+export async function getAllInternalTables(db?: Database): Promise<Table[]> {
+  if (!db) {
+    db = context.getAppDB()
+  }
+  const internalTables = await db.allDocs(
+    getTableParams(null, {
+      include_docs: true,
+    })
+  )
+  return internalTables.rows.map((tableDoc: any) => ({
+    ...tableDoc.doc,
+    type: "internal",
+    sourceId: tableDoc.doc.sourceId || BudibaseInternalDB._id,
+  }))
+}
+
+async function getAllExternalTables(): Promise<Table[]> {
+  const datasources = await sdk.datasources.fetch({ enriched: true })
+  const allEntities = datasources.map(datasource => datasource.entities)
+  let final: Table[] = []
+  for (let entities of allEntities) {
+    if (entities) {
+      final = final.concat(Object.values(entities))
+    }
+  }
+  return final
+}
+
+export async function getExternalTable(
+  datasourceId: string,
+  tableName: string
+): Promise<Table> {
+  const entities = await getExternalTablesInDatasource(datasourceId)
+  return entities[tableName]
+}
+
+export async function getTable(tableId: any): Promise<Table> {
+  const db = context.getAppDB()
+  if (isExternalTable(tableId)) {
+    let { datasourceId, tableName } = breakExternalTableId(tableId)
+    const datasource = await datasources.get(datasourceId!)
+    const table = await getExternalTable(datasourceId!, tableName!)
+    return { ...table, sql: isSQL(datasource) }
+  } else {
+    return db.get(tableId)
+  }
+}
+
+export async function getAllTables() {
+  const [internal, external] = await Promise.all([
+    getAllInternalTables(),
+    getAllExternalTables(),
+  ])
+  return [...internal, external]
+}
+
+export async function getExternalTablesInDatasource(
+  datasourceId: string
+): Promise<Record<string, Table>> {
+  const datasource = await datasources.get(datasourceId, { enriched: true })
+  if (!datasource || !datasource.entities) {
+    throw "Datasource is not configured fully."
```

Comment:
When would this happen in practice, and what would you be expected to do about it?

Reply:
This is very unlikely to happen, but it is likely there to protect against some sort of very old app version, where a datasource has been mal-configured. This shouldn't happen in modern versions, but apps can be imported from very old versions and datasources could be in a bad state - I don't think I've ever seen this error occur, but realistically the user would need to delete and recreate their datasource if it were to happen to get into such a bad state.